### PR TITLE
AWS/Route53Zone - create private hosted zone associated with VPC.

### PR DESCRIPTION
**Pull request by johnrengelman**
Add ability to create a private Hosted Zone that is associated to a VPC. 
Currently only supports a single VPC at creation time. 

Future question is how to support multiple VPCs. Amazon uses a Associate/Disassociate request, however this is problematic since you must specify one (and only one) VPC during creation.

Alternatively, we could model it as a VPC Set and and use the first entry as the entry for creating and then associate the others.

### CODE REVIEW

**Comment by johnrengelman** [comment]
Provides support for https://github.com/hashicorp/terraform/issues/1503

**Comment by johnrengelman** [comment]
This is missing pieces for read/update to sync the VPC list for the hosted zone. Probably need to answer the above question and add support before this gets merged.

**Comment by phinze** [comment]
Hey thanks for this!

> Future question is how to support multiple VPCs. Amazon uses a Associate/Disassociate request, however this is problematic since you must specify one (and only one) VPC during creation.

Our strategy with other similar scenarios has been to model the "attachment" or "association" as a resource itself. If I'm understanding the R53 API here correctly, this is a little awkward because you can specify a single VPC at zone creation time, but then N others later.

```js
aws_vpc "foo" {
  cidr_block = "10.0.0.0/16"
}
aws_vpc "bar" {
  cidr_block = "10.1.0.0/16"
}
aws_route53_zone "main" {
  vpc_id = "${aws_vpc.foo.id}"
  vpc_region = "us-east-1"
}
/* proposed additional resource */
aws_route53_zone_vpc_association "main-bar" {
  zone_id = "${aws_route_53_zone.main.id}"
  vpc_id = "${aws_vpc.bar.id}"
}
```

Having a top-level resource makes the CRUD operations very straightforward:

 * _Create_: `AssociateVPCWithHostedZone`
 * _Delete_: `DisassociateVPCWithHostedZone`
 * _Read_: `GetHostedZone` and walk through `VPCs`

For a minute I thought we could just ignore the `VPC` fields during zone creation so we could model everything with the separate resources, but alas, those fields are what trigger a private zone vs a public one. :disappointed: 

So the sadness is that because of this asymmetry in the R53 API, the VPC you specify in the Zone fields will always be the "root" VPC for the zone from Terraform's perspective.

Let me know if this all makes sense to you, and what your take is.

Thanks again for putting the time and effort in here! :grinning:

**Comment by johnrengelman** [comment]
I was thinking it would be easier to represent it like this:
```
aws_vpc "foo" {
  cidr_block = "10.0.0.0/16"
}
aws_vpc "bar" {
  cidr_block = "10.1.0.0/16"
}
aws_route53_zone "main" {
  vpc {
    id = "${aws_vpc.foo.id}"
  }
  vpc {
    id = "${aws_vpc.bar.id}"
    region = "us-east-1"
  }
}
```

as for the "zone_association" part...this is especially useful if the route53 zone is created outside of Terraform and you are just trying to attach your VPC to it. Would it be possibly to do something like collapse the `zone_association` into the `aws_route53_zone` resource (as a `vpc` entry) if it exists in the graph, otherwise treat it as a standalone resource?

**Comment by johnrengelman** [comment]
Ha, I just realized there is an existing PR adding this same support (https://github.com/hashicorp/terraform/pull/1159). I've cleaned up some of this code based on that. 
However both PRs have the same issue in that they don't correctly handle the VPC association on a read. I think the only way to do this is (Based on the Route53 API - http://docs.aws.amazon.com/Route53/latest/APIReference/API-get-hosted-zone-private.html) is to model the VPC associations on the HostedZone resource itself.

Would it be possible to collapse a `resource "aws_route53_hosted_zone_assocation"` into a `resource "aws_route53_hosted_zone"` if it exists in Terraform? 

In other words:
```
resource "aws_route53_hosted_zone"  "test"{
   name ="test.com"
   vpc {
      id = "123abc"
   }
}

resource "aws_route53_hosted_zone_association" "foo" {
   hosted_zone = "${aws_route53_hosted_zone.test.id"
   vpc_id = "456xyz"
}
```

would be equivalent to
```
resource "aws_route53_hosted_zone" "test" {
  name = "test"
  vpc {
    id = "123abc"
  }
  vpc {
    id = "456xyz"
  }
}
```
?

**Comment by c4milo** [comment]
It seems we have two contending PRs for this. Which one is it going to be merged?

**Comment by johnrengelman** [comment]
The problem is that there is an outstanding question about how to implement this. The route53 API in this area doesn't map well to Terraform.

**Comment by c4milo** [comment]
To be consistent with Route53 API, I would personally prefer: 

```
resource "aws_route53_hosted_zone"  "test"{
   name ="test.com"
   vpc {
      id = "123abc"
   }
}

resource "aws_route53_hosted_zone_association" "foo" {
   hosted_zone = "${aws_route53_hosted_zone.test.id"
   vpc_id = "456xyz"
}
```

**Comment by johnrengelman** [comment]
Unfortunately that is not consistent with the Route53 APIs because all associated VPCs are returned on the DescribePrivateHostedZone API call. So it would be difficult to distinguish which one was attached as part of HostedZone creation and which as an association.

It's legal to do this in Route53:
1) Create HostedZone "test" w/ VPC "123abc"
2) Associate VPC "456xyz" to HostedZone "test"
3) Deassociate VPC "123abc" from HostedZone "test"

There's not notion in AWS which VPC was used to create the Private HostedZone.

**Comment by c4milo** [comment]
oh I see, that's a bummer!. Thanks for explaining further @johnrengelman.

**Comment by jonhatalla** [comment]
Perhaps limiting the association of a private hosted zone to a single VPC might mitigate 'what's legal in route53' versus 'what's maintainable/consise in terraform'.  
In your situation, this would be the 'terraform process...
1) Create HostedZone "test" w/ VPC "123abc"
2) Create new HostedZone "test" w/VPC "456xyz"
3) Deassociate VPC "123abc" from first HostedZone "test"
4) delete first hosted zone 'test'

**Comment by johnrengelman** [comment]
I guess we could model everything as an association instead of a hosted_zone and associations.
I think there is still a problem here because what if you are not managing the hosted zone with Terraform but you want to manage the association? You could then potentially delete the hosted zone even though terraform isn't managing it (since removing the last association would in effect delete the zone). I still think the only way this can work well is if the association code can check if the hosted zone itself is being managed by terraform. That way if the hosted zone is not managed by Terraform then you treat the associations as individual items, otherwise it collapses everything into a single HostedZone resource and manage it from there.

**Comment by jonhatalla** [comment]
Yeah, Totally agree that everything should be managed by with Terraform (creation, destruction, associations).

**Comment by c4milo** [comment]
@johnrengelman I would do something similar to what @phinze suggested. Using AWS's default VPC for the initial creation of the zone, then swapping that with associations. I did something similar for `DHCP Options Set` but upon removing association: https://github.com/hashicorp/terraform/pull/1721

**Comment by catsby** [comment]
Hey @johnrengelman – have you been able to review this thread and given any thought to breaking the association out into it's own resource? We feel that's the best way forward here.

**Comment by johnrengelman** [comment]
Yeah. I think that's right way to go. I'll try and re-work this PR in the next few days.

**Comment by kendawg2** [comment]
This runs the plan OK, but on apply it always fails somewhere in the creation process.   It actually creates the zone as a private zone as expected, but throws an error and never adds the RRs to the zone.   

aws_route53_zone.primary: Error: 1 error(s) occurred:

* unexpected EOF
Error applying plan:

1 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* unexpected EOF

And then pretty much dumps after that.    A destroy does not remove the zone as expected.

I know this is WIP so this may be expected.

**Comment by kendawg2** [comment]
@johnrengelman @catsby @phinze Not trying to add complexity here, but the basic functionality of being able to create a simple private hosted zone for a particular plan is probably the 99% use case.   Being able to later associate an existing private zone is something needed but I would view as additive and a separate effort (one appears underway #1827.   I believe these are complimentary efforts which can stand alone.   I really don't like using the default VPC for initial association because you will have to disassociate it immediately so others can be created later.   Particularly with DHCP options.   It also can't cleanly destroy without impacting other VPC's a particular plan has no knowledge of.   Just my $.02 worth.   

Simply adding a VPC and region should automatically create a private vs public zone.   If it isn't created with a VPC it will be public and there is no way to make it private after that.   So when the zone is created, it needs to have a VPC associated.   #1159 addressed this fairly elegantly as does this code (they are similar).   #1827 extends that to add an already existing private zone to another VPC (or at least should do that).

**Comment by johnrengelman** [comment]
@kendawg2 Agreed. I didn't plan on using the default VPC because there are couple issues around that: 1) how do you identify the "default" VPC, 2) What do you do if someone deletes the "default" VPC.

I agree the association support can come after this since they are building on top of this PR.
In that case I'll run the tests and work out the bugs there.

**Comment by pmoust** [comment]
@johnrengelman I cherry picked your commit in #1827 (also changing the `vpc_region` check a bit).

**Comment by johnrengelman** [comment]
Yeah, I ran into the same issues with the `vpc_region` check and fix it. I'm rebasing my commits now and am going to add a couple more tests.
Would you be interested in just submitting your additions back to my branch and then we can bundle them up in 1 PR?

**Comment by pmoust** [comment]
I don't mind at all, feel free to pull. I just want this implemented, reviewed and merged so I can move along :>

**Comment by johnrengelman** [comment]
Great. I just rebased my commit and it looks like there are conflicts with the new nameserver changes that went in (of course the route53 API deviates here as well). Once I get those worked out and all my tests passing, then I'm pull your changes in.

**Comment by johnrengelman** [comment]
I've pulled in the `route53_hosted_zone_association` changes from here: https://github.com/hashicorp/terraform/pull/1827
The acceptance tests for both `route53_hosted_zone` and `route53_hosted_zone_association` now pass. 
@catsby - Mind taking a look?

**Comment by kendawg2** [comment]
@johnrengelman I merged this with a local master.   Ran it on a small and large (120 records) plan and it performed correctly.   Including destroy.   I did not test with tags and name servers but the private zone was correctly created, updated, and destroyed without errors.

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,8 +2,10 @@
 
 IMPROVEMENTS:
 
+  * **New resource: `aws_route53_zone_association`**
   * provider/aws: `aws_s3_bucket` exports `hosted_zone_id` and `region` [GH-1865]
   * provider/aws: `aws_route53_record` exports `fqdn` [GH-1847]
+  * provider/aws: `aws_route53_hosted_zone` can create private hosted zones [GH-1526]
```
Comment (pmoust) [inline CHANGELOG.md]: `aws_route53_zone`

**Comment by pmoust** [comment]
@johnrengelman no way to export the delegation set for a private hosted zone?
This is required when using a subdomain as private hosted zone where parent is a public zone.

**Comment by johnrengelman** [comment]
There's no API that exports those values for private zones.

File: builtin/providers/aws/resource_aws_route53_zone.go [hunk 2]
```diff
@@ -64,6 +86,7 @@ func resourceAwsRoute53ZoneCreate(d *schema.ResourceData, meta interface{}) erro
 	zone := cleanZoneID(*resp.HostedZone.ID)
 	d.Set("zone_id", zone)
 	d.SetId(zone)
+	d.Set("vpc_region", req.VPC.VPCRegion)
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_zone.go]: If a config does not specify `vpc_id`, this causes a panic

Comment (johnrengelman) [inline builtin/providers/aws/resource_aws_route53_zone.go]: ah yeah. We allow you to omit the `vpc_region` and then you default to using the region of the `aws` provider. In that case I need to make the field a `Calculated: true` field and explicitly set it when omitted. Otherwise I was running into diff mismatch errors during the acceptance tests. I guess this `d.Set` should be up in the `if` block above so it only occurs when using the `vpc` fields. Does that make sense?

**Comment by pmoust** [comment]
I have to insist a bit here regarding the NS records. Perhaps we should add an extra call for the NS record of the private zone's FQDN and add those four records under nameservers irregardless of this given in a manner different than public zones.
I do need to be able to access the `name_servers` attribute of a private zone within Terraform in order to update other zones in cases where private hosted zone is a subdomain of another zone, or other DNS provider.

**Comment by johnrengelman** [comment]
You're not understanding. Amazon does not have an API to retrieve the DelegationSet/Nameservers for a private hosted zone. It just doesn't exist as far as I can tell. If you can find it, I can add it: http://docs.aws.amazon.com/Route53/latest/APIReference/Welcome.html

File: builtin/providers/aws/resource_aws_route53_zone_association_test.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/terraform"
+
+	"github.com/awslabs/aws-sdk-go/aws"
+	"github.com/awslabs/aws-sdk-go/service/route53"
+)
+
+func TestAccRoute53ZoneAssociation(t *testing.T) {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_zone_association_test.go]: Could you please add a `_basic` suffix here? We're starting to adopt that scheme for tests, otherwise we can't run `TestAccRoute53ZoneAssociation` in isolation without also running `TestAccRoute53ZoneAssociationWithRegion`. `TestAccRoute53ZoneAssociationWithRegion` should then be `TestAccRoute53ZoneAssociation_WithRegion`.

**Comment by pmoust** [comment]
I do understand, 
http://docs.aws.amazon.com/Route53/latest/APIReference/API_ListResourceRecordSets.html#API_ListResourceRecordSets_Requests
Just get the resource record of `NS` type for the FQDN of the private zone.

File: builtin/providers/aws/resource_aws_route53_zone_association_test.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckRoute53ZoneAssociationExists("aws_route53_zone_association.foobar", &zone),
+				),
+			},
+		},
+	})
+}
+
+func TestAccRoute53ZoneAssociationWithRegion(t *testing.T) {
+	var zone route53.HostedZone
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_zone_association_test.go]: We pass this `zone` var around through almost all the functions below, but beyond setting it `*zone = *resp.HostedZone` I don't see it's actually used. Should we be checking a value on it? Otherwise, it seems to be baggage.

**Comment by johnrengelman** [comment]
@pmoust sorry, I stand corrected. For whatever reason, I dug through the API for an hour this morning trying to figure this out and I kept glancing right over that. 
I'll add that support.

**Comment by catsby** [comment]
I'm trying this out with the following config:

```javascript
provider "aws" {
   region = "us-west-2"
}

resource "aws_vpc" "test" {
  cidr_block = "10.0.0.0/16"
  tags {
    Name = "tf-r53-hz-test"
  }
}

resource "aws_route53_zone" "foo" {
  name = "bar.com"
  vpc_id = "${aws_vpc.test.id}"
}

resource "aws_route53_zone_association" "foobar" {
	zone_id = "${aws_route53_zone.foo.id}"
	vpc_id  = "${aws_vpc.test.id}"
}
```

I'm getting this error:

    * ConflictingDomainExists: Cannot associate with private zone: bar.com. as vpc vpc-2c67e449 in region us-west-2 is already associated with domain bar.com.

Am I doing it wrong?

File: builtin/providers/aws/resource_aws_route53_zone.go (hunk 2 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_zone.go]: > I guess this d.Set should be up in the if block above so it only occurs when using the vpc fields. Does that make sense?

It may make more sense to set it in the `Read` method

**Comment by johnrengelman** [comment]
Yes you are. When you create the `aws_route53_zone` with the `vpc_id`, you are implicitly associating the VPC to the zone. This is required by the Route53 API to create a private hosted zone. Thereforce, the `aws_route53_zone_association` is not needed. Thus the error, it's saying that VPC is already associated to the zone.
You would use the `zone_association` to associate another VPC to this private hosted zone that you just created.

**Comment by catsby** [comment]
Ah, I see that now in the `testAccRoute53ZoneAssociationRegionConfig` config, thanks.

**Comment by johnrengelman** [comment]
Updated this PR with the following:
* Comments from @catsby 
* Retrieve nameservers for private hosted zones (I haven't run the acceptance tests with this change, I might get to that this afternoon, will update here if I do).

**Comment by mzupan** [comment]
I just tested the PR pretty heavily creating/edit/deleting private zones and it works great

:+1:

**Comment by kendawg2** [comment]
Tested as well.

:+1:

**Comment by catsby** [comment]
Need a link placed in [layouts/aws.eb](https://github.com/hashicorp/terraform/blob/master/website/source/layouts/aws.erb)

**Comment by johnrengelman** [comment]
added sidebar item to `layouts/aws.erb` and rebased to current master.

**Comment by pmoust** [comment]
Thanks for the good work, seems great.

**Comment by gposton** [comment]
anxiously awaiting this! @KFishner

**Comment by johnrengelman** [comment]
Rebased to latest master. Hopefully this can get merged?

**Comment by catsby** [comment]
This won't merge cleanly, so I'm going to pull it down locally and poke at it. 
If I have changes that I can't figure out, I'll have to open a new PR. Stay tuned!

**Comment by johnrengelman** [comment]
I can rebase quick if you want me to. I'm sure it's Changelog.

**Comment by johnrengelman** [comment]
@catsby I just rebased this to the current master. It was just the Changelog that was conflicting.

**Comment by phinze** [comment]
@johnrengelman ah yeah we usually keep changelog out of PRs for that reason - the core team generally takes care of the changelog post-merge :+1:

**Comment by johnrengelman** [comment]
Noted for future. Thanks.

**Comment by catsby** [comment]
It works!

vpcs


Thanks all for the hard work in designing and getting this working. Clicking the button now.

**Comment by saulshanabrook** [comment]
@johnrengelman To make a private route53 zone, can I just do something like this?

```
resource "aws_route53_zone" "private" {
  name = "${var.private_route_name}"
  vpc_id = "${aws_vpc.default.id}"
  tags {
    Version = "2"
  }
}
```

Or do I need an `aws_route53_zone_association`?

**Comment by kendawg2** [comment]
@saulshanabrook you only need the association after the zone is created in order to associate with additional VPC's.   Your example will work for creating it and associating it with that VPC.

**Comment by saulshanabrook** [comment]
@kendawg2 Thanks!

**Comment by CaptTofu** [comment]
Hi all, I know this is a year later, but how do I create two zones -- one for forwared dns records (A, CNAME, etc) and one for reverse (PTR)

resource "aws_route53_zone" "primary" {
   name = "${var.aws_cluster_domain}"
   vpc_id = "${aws_vpc.vpc.id}"
}

resource "aws_route53_zone" "reverse" {
   name = "${var.aws_cluster_domain_reverse}"
   vpc_id = "${aws_vpc.vpc.id}"
}

This is for the same VPC,but it doesn't work.

In the UI, I have to create a zone, private, I can set a VPC
Then, I create another private zone, but not set the VPC, but after it's created, associate it.

I can't force the zone to be private unless I set a vpc_id, so I have a bit of a chicken-and-egg
